compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/TypeUtils.kt

```diff
@@ -218,14 +225,24 @@ fun FirTypeRef.withReplacedConeType(
     require(this is FirResolvedTypeRef)
     if (newType == null) return this
 
-    return buildResolvedTypeRef {
-        source = if (firFakeSourceElementKind != null)
-            this@withReplacedConeType.source?.fakeElement(firFakeSourceElementKind)
+    val newSource =
+        if (firFakeSourceElementKind != null)
+            this.source?.fakeElement(firFakeSourceElementKind)
         else
-            this@withReplacedConeType.source
-        type = newType
-        annotations += this@withReplacedConeType.annotations
-        delegatedTypeRef = this@withReplacedConeType.delegatedTypeRef
+            this.source
+
+    return if (newType is ConeKotlinErrorType) {
```

Comment:
... and may be also here (with some change of original function)

Reply:
... and this one is not trivial, since this util provides a way to alter `source`. Once we call `resolvedTypeFromPrototype` and get `FirResolvedTypeRef`, we can't replace `source` on it? So, I wonder it's better to use builders directly here.